### ISSUE TITLE
lxc publish and lxc image export creates lxd_lxd_metadata_id in archive root instead metadata.yaml

### ISSUE THREAD

**Issue by cryptofuture** [lxc/lxd#2467]
lxc publish and lxc image export creates lxd_lxd_metadata_id in archive root instead metadata.yaml. Renaming lxd_lxd_metadata_id to metadata.yaml in archive root helps. 
Without renaming import shows this error:
`error: Could not extract image metadata.yaml from tar: exit status 2 (tar: metadata.yaml: Not found in archive)`

```
  driver: lxc
  driverversion: 2.0.4
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 4.4.0-41-generic
  server: lxd
  serverpid: 820
  serverversion: 2.0.4
  storage: dir
  storageversion: ""
config:
  core.trust_password: true
public: false

```

**Comment by stgraber** [comment]
That one seems a bit odd given that the testsuite does do a launch => publish => export => import => launch. I'll try again manually to see if I can reproduce what you're seeing.

**Comment by stgraber** [comment]
Unable to reproduce here, using a similar daemon setup:
```
root@xenial:~# lxc info
apiextensions: []
apistatus: stable
apiversion: "1.0"
auth: trusted
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIFwDCCA6igAwIBAgIQCN4TB1F4aAXBsNZZzHeFgzANBgkqhkiG9w0BAQsFADA0
    MRwwGgYDVQQKExNsaW51eGNvbnRhaW5lcnMub3JnMRQwEgYDVQQDDAtyb290QHhl
    bmlhbDAeFw0xNjEwMDkwMTI0MDhaFw0yNjEwMDcwMTI0MDhaMDQxHDAaBgNVBAoT
    E2xpbnV4Y29udGFpbmVycy5vcmcxFDASBgNVBAMMC3Jvb3RAeGVuaWFsMIICIjAN
    BgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA2vW5zu+Yd0C/7wmSZs57jMKiex80
    vQ7cw38TXDy5wrhWg1uNRg6mCw2LPx1QuTgNvaLsqWtxHby4L5YF/OQTGEzra6wh
    u3EOTHDRWaE2+JRvuaNkHChTdLfDPJptBTTYWV1uLnb2c3CRcdl26TlCZzi5XN3W
    b3LbNkVKVy0V9wHO51HSX6Ph2Ggn1gyjiW7foEITrH2C1DybD0DUNwjy/hb9xFKV
    a8r/tQlqZJdcXKqzarAbsOTw1iPzpw/2PHgLKT0o0sa3XTmBU+UZLDkHxjw1wQM/
    J9vS3X8glGu0nmV/HNAXkWIWj5cDDrBFjfvW/lvsGfUEKu1lJIpUwwkVBiWaPnPn
    UppNYuVWHrdTLXch7ubvlBPFjl7tznfDtMh3cO20t4Lpf1Ra0RH1Gfs1JD3D00Vb
    kRnucuNfiJG4Ymr04RLiuMwhe7hKNZv485IzsiX3R8bKmKd8M3z6538oCNNUOYiC
    fqXBD2lMcmw3rDqwS5sdaXHp/PBOQec2YIGaLHj5tES/6uTElKSQtzrb5rcWpEKN
    Ah999csHtQf7hchQljhWyWs5qX38cxgB7iigVIM5Qxh9SExrUNSuwyLdJrwp32nJ
    GVpnfZveyloE6hlwq3RHRXaLNeJ7bvQpT1BUnoMxpQ0HyLqCFYDOFqRqw1lnQs6v
    GfFpQwD4HU+7WkMCAwEAAaOBzTCByjAOBgNVHQ8BAf8EBAMCBaAwEwYDVR0lBAww
    CgYIKwYBBQUHAwEwDAYDVR0TAQH/BAIwADCBlAYDVR0RBIGMMIGJggZ4ZW5pYWyC
    G2ZlODA6OmE4NWQ6MmZmOmZlMjA6MzRhYi82NIIKZmU4MDo6MS82NIIPMTAuMTU2
    LjAuMjE0LzI0gihmZDE6Njk5Nzo0OTM5OjQ5NWQ6MjE2OjNlZmY6ZmUyMToyZWNh
    LzY0ghtmZTgwOjoyMTY6M2VmZjpmZTIxOjJlY2EvNjQwDQYJKoZIhvcNAQELBQAD
    ggIBAL9+H2xiqdTpLrucHjVE/tPMPF1wRdGowhppOdwwocvKewa7E6IwEwIVXMu7
    dqwqK0rcH9LifBz9HFyidrhprDNdz/KhUmw971W+q7bxTAQiAZi8eM9lAB4RQq81
    R3+oVk98ZRV1NlGjLg2vdg3RxCx7TMTl7IVN8A13mXwb3z15we0rl3+q6WgbsqA/
    xB7gORhvgRLCZ/BoAIvIta8+uf4sqJzrpLdjiaV5panh9R8STdzlxbjDNleOp4UN
    Ob5gXtWVlIRrldeMk/PaD0H6LMVz1GySEQItzXXjJxD/c0R4R/AcCsXMVOxH1sKK
    NCOZokiSyb+PqtxcWR+Ifs73F2KG8bikq2XtIBozdfSRZADzSAxyiDdYUOweHchy
    u1th34EiopvdzNbMhFINGu0HsDC/dBJJEE4rICkvJrSvS9lVx9XsGvdL2FQIN664
    a5nMtF+i8URgG0jVW7F1fnVy0oX1BtIETy9eUNYPl/nM00a04Z/z8P02QEZsIsS+
    NvrmVzMXaaVslZnDF1+Z1rYt4I4ypFXT3LBvNez/36oesXz9PKSj9OnK3SATEcnf
    rKfYWSUw0UzjvTFSwyaw5RXl9QaPSBXXMB7hMyudeFYmIuxWr7Lmrp02wMGl3WkT
    Bjgo6O57AsyS53VctjNs4THOPXEA8JHbGQWetLURvMYBUPqT
    -----END CERTIFICATE-----
  certificatefingerprint: 4b2a5ce0a9d320d32a42db680ccd71c89db0e040b0bd6d04d0739fb1718a25c4
  driver: lxc
  driverversion: 2.0.4
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 4.4.0-38-generic
  server: lxd
  serverpid: 6336
  serverversion: 2.0.4
  storage: dir
  storageversion: ""
config: {}
public: false
```

```
root@xenial:~# lxc launch images:alpine/edge c1
Creating c1
Retrieving image: 100%
Starting c1
root@xenial:~# lxc stop c1
root@xenial:~# lxc publish c1
Container published with fingerprint: 15453bb0f56c2f8f7af8436ecaf697d1e0440820baa121052246c47b8633c884
root@xenial:~# lxc image export 154
Output is in 154.tar.gz
root@xenial:~# lxc image delete 154
root@xenial:~# lxc image import 154.tar.gz
Transferring image: 100%
Image imported with fingerprint: 15453bb0f56c2f8f7af8436ecaf697d1e0440820baa121052246c47b8633c884
root@xenial:~# lxc launch 1545 c2
Creating c2
Starting c2
root@xenial:~# lxc list
+------+---------+----------------------+-----------------------------------------------+------------+-----------+
| NAME |  STATE  |         IPV4         |                     IPV6                      |    TYPE    | SNAPSHOTS |
+------+---------+----------------------+-----------------------------------------------+------------+-----------+
| c1   | STOPPED |                      |                                               | PERSISTENT | 0         |
+------+---------+----------------------+-----------------------------------------------+------------+-----------+
| c2   | RUNNING | 10.85.216.234 (eth0) | fd30:630c:6171:6d86:216:3eff:fe0b:e3fd (eth0) | PERSISTENT | 0         |
+------+---------+----------------------+-----------------------------------------------+------------+-----------+
```

**Comment by stgraber** [comment]
And I confirmed that the tarball does contain "metadata.yaml".

**Comment by cryptofuture** [comment]
Finally was able to reproduce it again. Like that:
```
container1 - stopped
lxc copy container1 test
lxc publish test
lxc image export xxx
```
Archive contains lxd_lxd_metadata_id in acrchive root.

**Comment by stgraber** [comment]
Ah, that may then be explained by a bug we fixed in 2.0.5 where metadata.yaml wouldn't be copied when copying containers using the directory backend.

**Comment by stgraber** [comment]
Confirmed that LXD 2.4 doesn't suffer from this so neither will 2.0.5.

That being said, the fact that a metadata file is generated but with a wrong file name is still a bug we should fix. Looking into this now.

**Comment by cryptofuture** [comment]
>Ah, that may then be explained by a bug we fixed in 2.0.5
k, it already fixed in upstream than.
Btw, in lxd from lxd-git-master same containers gives EOF on publishing. But yes it was created in 2.0.4 or earlier (in 2.0.x for sure) and a lot things has changed since it.
As for 2.0.4 with directory backend it gives lxd_lxd_metadata_id, only when publishing container, created as copy from another container, just right after the copy.

Trying the same with btrfs backend on 2.0.4 right now.

Another interesting fact, non related to this issue. I use three different backends.  And even was able to test speed in all possible backends and plain on xfs/ext4 as separated partition and as on / on same hardware spec. Became btrfs user with lxd help.
Became btrfs user with lxd.

**Comment by stgraber** [comment]
Right, upgrading LXD won't fix this problem as the "fix" we have for it right now would require you to do the "lxc copy" step again.

The fix I pushed to my branch will fix the issue with existing containers that were created with older LXD versions.
